refactor!: improve Rule argument handling code

spflueger opened the pull request:
Extract the Rule argument handling code into a separate module. Also allow more flexible arguments and improve type annotation handling via inspect. Exchange the Rule interface with a `Protocol` (`ConservationRule`) and define a second type of Rule (`EdgeQNConservationRule`), which does not require any node quantum numbers.

This prepares things for issue #303

redeboer commented on expertsystem/reaction/conservation_rules.py:
```diff
@@ -229,7 +237,7 @@ class CParityNodeInput:
 
 
-class CParityConservation(Rule):
+class CParityConservation:
```
So here it's important to be aware that the developer is not warned  anymore (through `mypy` or other linters) within this module when a signature is incorrect. If I understand correctly, the `Protocol` is only enforced statically at the place where the rule is called, wereas with an ABC (`Rule`, previously), the signature is checked at the place of the definition.

redeboer commented on expertsystem/reaction/conservation_rules.py:
```diff
@@ -735,26 +746,25 @@ def __call__(
-        |L - S| \leq J \leq |L + S|
-    """
-
+class SpinMagnitudeConservation:
```
Yes, that's indeed better naming

redeboer commented on expertsystem/reaction/solving.py:
```diff
@@ -83,14 +91,37 @@ class GraphSettings:
     node_settings: Dict[int, NodeSettings] = attr.ib(factory=dict)
 
 
+# a tuple of rules defines a group which together violates all possible
```
Why a comment, not a docstring?

redeboer commented on expertsystem/reaction/solving.py:
```diff
@@ -83,14 +91,37 @@ class GraphSettings:
+# combinations that were processed
+def convert_to_names(
+    rules: Dict[int, Set[Tuple[Rule, ...]]]
+) -> Dict[int, Set[Tuple[str, ...]]]:
```
Why convert to a `str`?

spflueger replied:
This is related to the issue of uniquely identifying a rule. The user is only interested why a problem does not have any solutions, so for example because of c parity violation. A string is enough for this. Currently the rule objects are used themselves, and define a human readable str (repr) method as well as a hash function. This is not very transparent.

redeboer commented on expertsystem/reaction/solving.py:
```diff
@@ -746,16 +562,15 @@ def get_rules_by_priority(
+        arg_handler = RuleArgumentHandler()
+
         for node_id in self.__graph.nodes:
-            # currently we only have rules related to graph nodes
```
Thinking about this comment. Did it become outdated because of the new distinction between `EdgeInteractionRule` and `InteractionRule`?

spflueger replied:
Partly, especially the new upcoming edge rules will make it different

redeboer replied:
Upcoming edge rules?

redeboer commented on expertsystem/reaction/conservation_rules.py:
```diff
@@ -297,7 +305,7 @@ class GParityNodeInput:
 
 
-class GParityConservation(Rule):
+class GParityConservation:
```
Wait, why do we still use classes now that there are no abstract base classes anymore? There was some problem with the signatures or so, right?

(Can also be handled in a follow-up issue)

spflueger replied:
This is related to the fact that I was not able to get the Protocols working for regular functions.

spflueger replied:
Since I now know that the full signature of the protocol functions is important, also the argument names, I can verify that it also works with regular functions. So this can be changed

redeboer commented on tests/unit/reaction/conservation_rules/test_additive_qn.py:
```diff
@@ -8,6 +8,4 @@
-    rule = ChargeConservation()
-
-    assert rule(*graph_input) == expected_value
+    assert ChargeConservation()(*graph_input) == expected_value
```
See question on why the rules are still classes.

spflueger replied:
The rules created with the additive quantum number decorator have the additional difficulty to achieve the pickling support. With classes this much easier to sidestep

redeboer replied:
Ah yes, the pickling was the problem

redeboer reviewed (commented):
I added some thoughts inline. Overall I think that `Protocol` (static duck typing [PEP 544](https://www.python.org/dev/peps/pep-0544)) is the way to go here. I'm still not 100% convinced though, because the typing takes place outside the module where the rules are defined. But it does seem that `Protocol` offers some more flexibility (I'm for instance also thinking about `kwargs`, but they are not needed at this stage).

Note: as discussed earlier, the strong emphasis on typing we introduced over the passed releases, seems to have caused quite some pain within the `conservation_rules` and `solving` modules. That's what I mean with flexibility, that I hope that `Protocol` brings back some of the duck typing behaviour that we used to have, but still with the possibility to type as much as possible.

spflueger commented on expertsystem/reaction/conservation_rules.py:
```diff
@@ -1,20 +1,54 @@
+A rule is a function that takes quantum numbers as input and outputs a boolean.
+There are three different types of rules:
+
+1. Rules that work on individual graph edges.
```
yeah, i did not spend much effort on the doc strings yet as its still a bit experimental PR

spflueger commented:
> I added some thoughts inline. Overall I think that `Protocol` (static duck typing [PEP 544](https://www.python.org/dev/peps/pep-0544)) is the way to go here. I'm still not 100% convinced though, because the typing takes place outside the module where the rules are defined. But it does seem that `Protocol` offers some more flexibility (I'm for instance also thinking about `kwargs`, but they are not needed at this stage).
> 
> Note: as discussed earlier, the strong emphasis on typing we introduced over the passed releases, seems to have caused quite some pain within the `conservation_rules` and `solving` modules. That's what I mean with flexibility, that I hope that `Protocol` brings back some of the duck typing behaviour that we used to have, but still with the possibility to type as much as possible.

Thanks for the review. I'm mainly just unhappy with the rather imprecise rule interface/protocols. A precise definition of a rule in terms of an interface is needed to run the solving process. However our need is also the rather flexible definition of rules. Rule arguments can vary from quantum numbers of a single edge, to quantum numbers from all edge going in and out of a node, as well as quantum numbers from the node itself. Also it would be preferable if a rule can chose the names for arguments and they are not predetermined by the interface.
So the concept of an interface is not very suitable here. However I currently do not have any good concept to handle the solving without an interface of a rule. The hard part is to feel the quantum numbers of a state transition graph to a rule. So knowing which quantum numbers a rule wants or needs in order to execute, and where to find those in the graph

redeboer reviewed (approved):
Okay, GJ!
The only thing I find a bit counter-intuitive is the naming and in that extend the type of objects. This becomes particularly clear here:

https://github.com/ComPWA/expertsystem/blob/754666d3721d8ddfb461db398cce0fc7183c526f/expertsystem/reaction/_default_settings.py#L50-L73

Of course, this just reflects the fact that a switch from `ABC` to `Protocol` allows for duck typing, so I guess it's a matter of getting used to.

spflueger commented:
> Okay, GJ!
> The only thing I find a bit counter-intuitive is the naming and in that extend the type of objects. This becomes particularly clear here:
> 
> https://github.com/ComPWA/expertsystem/blob/754666d3721d8ddfb461db398cce0fc7183c526f/expertsystem/reaction/_default_settings.py#L50-L73
> 
> Of course, this just reflects the fact that a switch from `ABC` to `Protocol` allows for duck typing, so I guess it's a matter of getting used to.

Yes, I would also say a matter of getting used to. Lets see how this works for us, we can always still modify it.
